interpretation/src/main/kotlin/com/hartwig/actin/treatment/TreatmentRankingModel.kt

```diff
@@ -0,0 +1,85 @@
+import com.hartwig.actin.datamodel.molecular.driver.Fusion
+import com.hartwig.actin.datamodel.molecular.driver.GeneAlteration
+import com.hartwig.actin.datamodel.molecular.driver.Virus
+import com.hartwig.actin.datamodel.molecular.evidence.Actionable
+import com.hartwig.actin.datamodel.molecular.evidence.TreatmentEvidence
+import kotlin.math.exp
+
+data class TreatmentRankResult(val treatment: String, val scores: List<EvidenceScore>) : Comparable<TreatmentRankResult> {
+
+    val score = scores.sumOf { it.score }
+
+    override fun compareTo(other: TreatmentRankResult): Int {
+        return Comparator.comparingDouble<TreatmentRankResult> { it.score }.reversed().compare(this, other)
+    }
+}
+
+data class TreatmentEvidenceWithTarget(val treatmentEvidence: TreatmentEvidence, val target: String)
+data class DuplicateEvidenceGrouping(val treatment: String, val gene: String?, val tumorMatch: TumorMatch, val benefit: Boolean)
+
+class TreatmentRankingModel(private val scoringModel: EvidenceScoringModel) {
+
+    fun rank(record: PatientRecord): TreatmentEvidenceRanking {
+        val actionables = record.molecularHistory.molecularTests.asSequence().flatMap {
+            (it.drivers.fusions + it.drivers.variants + it.drivers.copyNumbers + it.drivers.homozygousDisruptions + it.drivers.disruptions + it.drivers.viruses) +
+                    it.characteristics.microsatelliteStability + it.characteristics.homologousRecombination + it.characteristics.tumorMutationalLoad + it.characteristics.tumorMutationalBurden
+        }.filterNotNull()
+
+        val treatmentEvidencesWithTarget = treatmentEvidencesWithTargets(actionables)
+        val scoredTreatments = treatmentEvidencesWithTarget.map { it to scoringModel.score(it.treatmentEvidence) }
+        val scoredTreatmentsWithDuplicatesDiminished = groupEvidenceForDuplicationAndDiminishScores(scoredTreatments)
+
+        val rankingResults = scoredTreatmentsWithDuplicatesDiminished.map { it.key.treatment to it.value }.groupBy { it.first }
+            .map { TreatmentRankResult(it.key, it.value.flatMap { t -> t.second }) }
+
+        return TreatmentEvidenceRanking(rankingResults.sorted().map {
+            RankedTreatment(it.treatment, it.scores.map { s -> s.event }.toSet(),
+                it.scores.sumOf { s -> s.score })
+        })
+    }
+
+    private fun groupEvidenceForDuplicationAndDiminishScores(scoredTreatmentEntries: Sequence<Pair<TreatmentEvidenceWithTarget, EvidenceScore>>) =
+        scoredTreatmentEntries.groupBy {
+            DuplicateEvidenceGrouping(
+                it.first.treatmentEvidence.treatment,
+                it.first.target,
+                it.second.scoringMatch.tumorMatch,
+                (it.second.score > 0)
+            )
+        }.mapValues { entry ->
+            saturatingDiminishingReturnsScore(entry.value.map { it.second })
+        }
+
+    private fun treatmentEvidencesWithTargets(actionables: Sequence<Actionable>) =
```

Comment:
@kduyvesteyn wondering what you think of this. The idea is that we want to group together evidences that are close enough conceptually to be considered duplicates, and that is why gene is getting passed around. However the previous implementation would not have worked properly for characteristics or fusions (they would have been treated as dups).  Also using gene is a bit of a shortcut as we don't know if two variants are part the same CKB "tree"

Then again, this is just an internal logic for an algo we already know is far from perfect, so maybe fine to move forward for now and not spend too much time on this until we have more data on the performance of the algo.

Reply:
This looks quite ok to me, maybe I'd throw on the else part in this function but that's a detail.

I have started growing more fond of the framework of the evidence ranking last few weeks and when reviewing this PR, there is probably not much more we can do given the data we have (ideally you'd weigh evidence based on # patients tested in the paper but we don't have that information). 

I can't imagine the actual weights we currently have are optimal but think there is a chance that the framework may hold once we actually have some retrospective data + truth to test against.

Reply:
In any case, the recent changes in grouping for diminishing scores look good to me!